cpp/src/experimental/random_walks.cuh

```diff
@@ -0,0 +1,854 @@
+#include <type_traits>
+
+namespace cugraph {
+namespace experimental {
+
+namespace detail {
+
+template <typename T>
+using device_vec_t = rmm::device_uvector<T>;
+
+template <typename T>
+using device_v_it = typename device_vec_t<T>::iterator;
+
+template <typename value_t>
+value_t* raw_ptr(device_vec_t<value_t>& dv)
+{
+  return dv.data();
+}
+
+template <typename value_t>
+value_t const* raw_const_ptr(device_vec_t<value_t> const& dv)
+{
+  return dv.data();
+}
+
+template <typename value_t, typename index_t = size_t>
+struct device_const_vector_view {
+  device_const_vector_view(value_t const* d_buffer, index_t size) : d_buffer_(d_buffer), size_(size)
+  {
+  }
+
+  device_const_vector_view(device_const_vector_view const& other) = delete;
+  device_const_vector_view& operator=(device_const_vector_view const& other) = delete;
+
+  device_const_vector_view(device_const_vector_view&& other)
+  {
+    d_buffer_ = other.d_buffer_;
+    size_     = other.size_;
+  }
+  device_const_vector_view& operator=(device_const_vector_view&& other)
+  {
+    d_buffer_ = other.d_buffer_;
+    size_     = other.size_;
+
+    return *this;
+  }
+
+  value_t const* begin(void) const { return d_buffer_; }
+
+  value_t const* end() const { return d_buffer_ + size_; }
+
+  index_t size(void) const { return size_; }
+
+ private:
+  value_t const* d_buffer_{nullptr};
+  index_t size_;
+};
+
+// raft random generator:
+// (using upper-bound cached "map"
+//  giving out_deg(v) for each v in [0, |V|);
+//  and a pre-generated vector of float random values
+//  in [0,1] to be brought into [0, d_ub[v]))
+//
+template <typename vertex_t,
+          typename edge_t,
+          typename seed_t  = long,
+          typename real_t  = float,
+          typename index_t = edge_t>
+struct rrandom_gen_t {
+  using seed_type = seed_t;
+  using real_type = real_t;
+
+  rrandom_gen_t(raft::handle_t const& handle,
+                index_t nPaths,
+                device_vec_t<real_t>& d_random,             // scratch-pad, non-coalesced
+                device_vec_t<edge_t> const& d_crt_out_deg,  // non-coalesced
+                seed_t seed = seed_t{})
+    : handle_(handle),
+      seed_(seed),
+      num_paths_(nPaths),
+      d_ptr_out_degs_(raw_const_ptr(d_crt_out_deg)),
+      d_ptr_random_(raw_ptr(d_random))
+  {
+    auto rnd_sz = d_random.size();
+
+    CUGRAPH_EXPECTS(rnd_sz >= static_cast<decltype(rnd_sz)>(nPaths), "Un-allocated random buffer.");
+
+    // done in constructor;
+    // this must be done at each step,
+    // but this object is constructed at each step;
+    //
+    raft::random::Rng rng(seed_);
+    rng.uniform<real_t, index_t>(
+      d_ptr_random_, nPaths, real_t{0.0}, real_t{1.0}, handle.get_stream());
+  }
+
+  // in place:
+  // for each v in [0, num_paths) {
+  // if out_deg(v) > 0
+  //   d_col_indx[v] = random index in [0, out_deg(v))
+  //}
+  void generate_col_indices(device_vec_t<vertex_t>& d_col_indx) const
+  {
+    thrust::transform_if(
```

Comment:
Aren't you using the return value of thrust::trasnform_if? The actual numbers of generated column indices can be smaller than num_paths_.

Reply:
Yes, the result of the `transform_if` is `d_col_indx`. And, yes, there can be fewer col indices than num_paths, which is why I used a `transform_if` in conjunction with `d_crt_out_degs`: for those vertices `v` with `out_deg(v) == 0`, corresponding entries in `d_col_indx` are _not_ generated.